File: packages/api-models-py/skyportal_py_models/reminders.py
"""Request and response models for ``/api/{resource_type}/{id}/reminders``."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from skyportal_py_models.groups import Group

ReminderResourceType = Literal[
    "source",
    "spectra",
    "gcn_event",
    "shift",
    "earthquake",
]


class Reminder(BaseModel):
    """A reminder on any remindable resource (upstream ``Reminder``).

    Upstream splits reminders across ``Reminder``, ``ReminderOnSpectrum``,
    ``ReminderOnGCN``, ``ReminderOnShift`` and ``ReminderOnEarthquake``;
    this model is the union of that family, so each type-specific foreign
    key is optional and only the ones belonging to the reminder's own
    table are ever set. ``user`` is the owner's ``User.to_dict()``, and
    ``obj``, ``spectrum``, ``gcn``, ``shift`` and ``earthquake`` stay
    ``dict`` to avoid importing in a circle from the modules that import
    this one.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime | None = None
    modified: datetime | None = None
    text: str | None = None
    origin: str | None = None
    bot: bool | None = None
    next_reminder: datetime | None = None
    reminder_delay: float | None = None
    number_of_reminders: int | None = None
    user_id: int | None = None
    user: dict[str, Any] | None = None
    groups: list[Group] | None = None
    obj_id: str | None = None
    spectrum_id: int | None = None
    gcn_id: int | None = None
    earthquake_id: int | None = None
    shift_id: int | None = None
    obj: dict[str, Any] | None = None
    spectrum: dict[str, Any] | None = None
    gcn: dict[str, Any] | None = None
    shift: dict[str, Any] | None = None
    earthquake: dict[str, Any] | None = None


class RemindersResponse(BaseModel):
    """All reminders attached to one resource."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    resource_id: str = Field(alias="resourceId")
    resource_type: str = Field(alias="resourceType")
    reminders: list[Reminder] = Field(default_factory=list)


class ReminderPost(BaseModel):
    """Payload for creating reminders on a resource."""

    model_config = ConfigDict(extra="forbid")

    text: str
    next_reminder: str
    reminder_delay: float | None = None
    number_of_reminders: int | None = None
    group_ids: list[int] | None = None
    user_ids: list[int] | None = None


class ReminderPostResponse(BaseModel):
    """IDs of the reminders created by a post."""

    model_config = ConfigDict(extra="forbid")

    reminder_ids: list[int] = Field(default_factory=list)


class ReminderUpdate(BaseModel):
    """Payload for updating an existing reminder."""

    model_config = ConfigDict(extra="forbid")

    text: str | None = None
    origin: str | None = None
    bot: bool | None = None
    next_reminder: str | None = None
    reminder_delay: float | None = None
    number_of_reminders: int | None = None
    group_ids: list[int] | None = None
    user_ids: list[int] | None = None

File: packages/api-models-ts/src/Reminders.ts
/**
 * Request and response models for `/api/{resourceType}/{id}/reminders`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Groups from "./Groups.ts";
import * as Schemas from "./Schemas.ts";

/**
 * What a reminder is attached to. Note the route uses the singular `"source"`.
 *
 * @since 1.0.0
 * @category Models
 */
export type ReminderResourceType =
    | "source"
    | "spectra"
    | "gcn_event"
    | "shift"
    | "earthquake";

/**
 * A reminder on any remindable resource (upstream `Reminder`).
 *
 * Upstream splits reminders across `Reminder`, `ReminderOnSpectrum`,
 * `ReminderOnGCN`, `ReminderOnShift` and `ReminderOnEarthquake`; this model is
 * the union of that family, so each type-specific foreign key is optional and
 * only the ones belonging to the reminder's own table are ever set. `user` is
 * the owner's `User.to_dict()`, and `obj`, `spectrum`, `gcn`, `shift` and
 * `earthquake` stay free-form to avoid importing in a circle from the modules
 * that import this one.
 *
 * @since 1.0.0
 * @category Models
 */
export const Reminder = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        text: Schemas.NullishString,
        origin: Schemas.NullishString,
        bot: Schemas.NullishBoolean,
        next_reminder: Schemas.NullishTimestamp,
        reminder_delay: Schemas.NullishNumber,
        number_of_reminders: Schemas.NullishInteger,
        user_id: Schemas.NullishInteger,
        user: Schemas.nullish(Schemas.JsonObject),
        groups: Schemas.nullish(v.array(Groups.Group)),
        obj_id: Schemas.NullishString,
        spectrum_id: Schemas.NullishInteger,
        gcn_id: Schemas.NullishInteger,
        earthquake_id: Schemas.NullishInteger,
        shift_id: Schemas.NullishInteger,
        obj: Schemas.nullish(Schemas.JsonObject),
        spectrum: Schemas.nullish(Schemas.JsonObject),
        gcn: Schemas.nullish(Schemas.JsonObject),
        shift: Schemas.nullish(Schemas.JsonObject),
        earthquake: Schemas.nullish(Schemas.JsonObject),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type Reminder = v.InferOutput<typeof Reminder>;

/**
 * All reminders attached to one resource.
 *
 * @since 1.0.0
 * @category Models
 */
export const RemindersResponse = Schemas.model(
    v.strictObject({
        resourceId: v.string(),
        resourceType: v.string(),
        reminders: Schemas.list(Reminder),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type RemindersResponse = v.InferOutput<typeof RemindersResponse>;

/**
 * Payload for creating reminders on a resource.
 *
 * If `user_ids` is omitted the reminder is created for the requesting user
 * only, and if `group_ids` is omitted it is visible to all of the requesting
 * user's groups. The server defaults `reminder_delay` and
 * `number_of_reminders` to `1`.
 *
 * @since 1.0.0
 * @category Models
 */
export interface ReminderPost {
    readonly text: string;
    readonly next_reminder: string;
    readonly reminder_delay?: number | undefined;
    readonly number_of_reminders?: number | undefined;
    readonly group_ids?: ReadonlyArray<number> | undefined;
    readonly user_ids?: ReadonlyArray<number> | undefined;
}

/**
 * IDs of the reminders created by a post.
 *
 * @since 1.0.0
 * @category Models
 */
export const ReminderPostResponse = Schemas.model(
    v.strictObject({
        reminder_ids: Schemas.list(Schemas.Integer),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type ReminderPostResponse = v.InferOutput<typeof ReminderPostResponse>;

/**
 * Payload for updating an existing reminder.
 *
 * @since 1.0.0
 * @category Models
 */
export interface ReminderUpdate {
    readonly text?: string | undefined;
    readonly origin?: string | undefined;
    readonly bot?: boolean | undefined;
    readonly next_reminder?: string | undefined;
    readonly reminder_delay?: number | undefined;
    readonly number_of_reminders?: number | undefined;
    readonly group_ids?: ReadonlyArray<number> | undefined;
    readonly user_ids?: ReadonlyArray<number> | undefined;
}
